torch_geometric/data/materialized_graph.py

```diff
@@ -0,0 +1,125 @@
+from abc import abstractmethod
+from collections.abc import MutableMapping
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Dict, Optional, Union
+
+from torch import LongTensor
+
+from torch_geometric.typing import EdgeTensorType, EdgeType
+from torch_geometric.utils.mixin import CastMixin
+
+
+class EdgeLayout(Enum):
+    COO = 'COO'
+    CSC = 'CSC'
+    CSR = 'CSR'
+    LIL = 'LIL'
+
+
+@dataclass
+class EdgeAttr(CastMixin):
+    r"""Defines the attributes of an :obj:`MaterializedGraph` edge."""
+
+    # The layout of the edge representation
+    layout: Optional[EdgeLayout] = None
+
+    # The type of the edge
+    edge_type: Optional[Any] = None
+
+
+class MaterializedGraph(MutableMapping):
+    def __init__(self, edge_attr_cls: Any = EdgeAttr):
+        r"""Initializes the materialized graph. Implementor classes can
+        customize the ordering and required nature of their :class:`EdgeAttr`
+        edge attributes by subclassing :class:`EdgeAttr` and passing the
+        subclass as :obj:`attr_cls`."""
+        super().__init__()
+        self.__dict__['_edge_attr_cls'] = edge_attr_cls
+
+    # Core ####################################################################
+
+    @abstractmethod
+    def _put_edge_index(self, edge_index: EdgeTensorType,
+                        edge_attr: EdgeAttr) -> bool:
+        return None
+
+    def put_edge_index(self, edge_index: EdgeTensorType, *args,
+                       **kwargs) -> bool:
+        r"""Synchronously adds an edge_index tensor to the materialized graph.
+
+        Args:
+            tensor(EdgeTensorType): an edge_index in a format specified in
+            attr.
+            **attr(EdgeAttr): the edge attributes.
+        """
+        edge_attr = self._edge_attr_cls.cast(*args, **kwargs)
+        if not edge_attr.layout:
+            raise AttributeError(
+                "An edge layout is required to store an edge index, but one "
+                "was not provided.")
+
+        # NOTE implementations should take care to ensure that `SparseTensor`
+        # objects are treated properly here.
+        return self._put_edge_index(edge_index,
+                                    self._edge_attr_cls.cast(*args, **kwargs))
+
+    @abstractmethod
+    def _get_edge_index(self, edge_attr: EdgeAttr) -> EdgeTensorType:
+        return None
+
+    def get_edge_index(self, *args, **kwargs) -> EdgeTensorType:
+        r"""Synchronously gets an edge_index tensor from the materialized
+        graph.
+
+        Args:
+            **attr(EdgeAttr): the edge attributes.
+
+        Returns:
+            EdgeTensorType: an edge_index tensor corresonding to the provided
+            attributes, or None if there is no such tensor.
+        """
+        return self._get_edge_index(self._edge_attr_cls.cast(*args, **kwargs))
+
+    def sample(
```

Comment:
I think we should add this in a later step once we have clarity on how to frame the the `pyg-lib` `graph` interface.

Reply:
That's fair. I currently included it just as a reference marker to indicate where we'd want to implement a method in the future, but I'm also happy to remove it since we won't have a corresponding implementation today. No strong preference, lmkwyt!

Reply:
Sounds good. Removed for now :)